packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -0,0 +1,342 @@
+import { assert, details, q } from '@agoric/assert';
+import { parseVatSlot } from '../parseVatSlots';
+
+export function makeVirtualObjectManager(
+  syscall,
+  allocateExportID,
+  valToSlotTable,
+  m,
+  cacheSize,
+) {
+  /**
+   * Make a simple LRU cache of virtual object inner selves.
+   *
+   * @param {number} size  Maximum number of entries to keep in the cache before
+   *    starting to throw them away.
+   */
+  function makeCache(size) {
+    let lruHead;
+    let lruTail;
+    let count = 0;
+    const liveTable = new Map();
+
+    const cache = {
+      makeRoom() {
+        while (count > size && lruTail) {
+          liveTable.delete(lruTail.instanceKey);
+          lruTail.dropData();
+          if (lruTail.prev) {
+            lruTail.prev.next = undefined;
+          } else {
+            lruHead = undefined;
+          }
+          lruTail = lruTail.prev;
+          count -= 1;
+        }
+      },
+      flush() {
+        const saveSize = size;
+        size = 0;
+        cache.makeRoom();
+        size = saveSize;
+      },
+      remember(innerObj) {
+        if (liveTable.has(innerObj.instanceKey)) {
```

Comment:
I'm a bit weirded out by the objects being passed into this structure being modified by it, I think I'd prefer the cache produce it's own boxes around the user-supplied values.

Reply:
I had a similar sense, but since the inner object and the cache are already necessarily deep in each other's business anyway there didn't seem to be much payoff to the added complexity.  Keep in mind that cache isn't dealing with "user-supplied values", it's only dealing with internal objects.